lib/Trace.js

```diff
@@ -0,0 +1,181 @@
+
+            const correlationId = this._type === 'Consumer' ? this._span.getParentId() : this._span.getId();
+            if (correlationId) {
+                node.correlationIds = [{
+                    value: correlationId,
+                    scope: 'Interaction',
+                }];
+            }
+        }
+
+        return node;
+    }
+
+    static _deriveOperation(span) {
+        const method = span.getTags()['http.method'];
+        if (method) {
+            return method;
+        }
+        return span.getOperationName();
+    }
+
+    static _deriveUrl(tags) {
+        let url = tags['http.url'];
+        if (url) {
+            return url;
+        }
+
+        url = tags['http.uri'];
+        if (url) {
+            return url;
+        }
+
+        url = tags['http.path'];
+        if (url) {
+            return url;
+        }
+
+        return null;
+    }
+
+    static _tagsToProperties(tags) {
+        const properties = [];
+        Object.keys(tags).forEach((objectKey) => {
+            properties.push({
+                name: objectKey,
+                value: tags[objectKey],
+                type: 'Text',
+            });
+        });
+        return properties;
+    }
+}
+
+class Trace {
+    constructor() {
+        this._nodes = [];
+    }
+
+    addNode(type, span, parentSpan) {
+        let nodes = this._nodes;
+
+        if (parentSpan && parentSpan.getId()) {
+            const node = this.findNode(parentSpan.getId());
+            if (node) {
+                nodes = node.getNodes();
+            }
+        }
+
+        // TODO throw exception if span is already there?
+        nodes.push(new Node(span, type));
+    }
+
+    setNodeType(type, span) {
+        const node = this.findNode(span.getId());
+        if (node) {
+            node.setType(type);
+        } else {
+            console.log('Error setNode node not found');
+        }
+    }
+
+    isParent(span) {
+        // we don't have to go deeper as the root spans are in top level array
+        for (let i = 0; i < this._nodes.length; i += 1) {
+            if (this._nodes[i].getSpan().getId() === span.getId()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    findNode(spanId) {
+        return Trace._findNode(spanId, this._nodes);
+    }
+
+    static _findNode(spanId, nodes) {
+        if (!nodes) {
+            return null;
+        }
+
+        for (let i = 0; i < nodes.length; i += 1) {
+            if (spanId === nodes[i].getSpan().getId()) {
+                return nodes[i];
+            }
+
+            const ret = Trace._findNode(spanId, nodes[i].getNodes());
+            if (ret) {
+                return ret;
+            }
+        }
+
+        return null;
+    }
+
+    fromSpan(span, principal) {
+        return {
+            traceId: span.getTraceId(),
+            fragmentId: span.getId(),
+            timestamp: span.getStartTime() * 1000,
+            principal,
```

Comment:
now it will be explicitly added by the application via the 'principal' tag - so it is up to them which spans they associated it with (and therefore which nodes it iis recorded on).